### ISSUE TITLE
Wrong code for avx2-i32x16.

### ISSUE THREAD

**Issue by Vsevolod-Livinskij** [ispc/ispc#1793]
ISPC generates wrong code for avx2-i32x16.

Error:
> \>$ ispc -O0 --target=avx2-i32x16 -o func.o func.ispc && clang++ func.o driver.cpp && ./a.out
0
> \>$ ispc -O2 --target=avx2-i32x16 -o func.o func.ispc && clang++ func.o driver.cpp && ./a.out
1

Reproducer:
func.ispc
```c++
extern uniform unsigned int var_2;
extern uniform bool var_14;
extern uniform int arr_29;
extern uniform unsigned int64 arr_68;

export void test() {
    arr_29 = none(var_2);
    if ((varying int8) var_14 < 0)
        arr_68 = 1;
}
```

driver.cpp
```c++
#include <stdio.h>

unsigned int var_2 = 3227406728U;
bool var_14 = 1;
int arr_29;
unsigned long long int arr_68;

extern "C" { void test(); }
int main() {
    test();
    printf("%llu\n", arr_68);
}
```

ISPC version:
 1.14.0dev (build commit 9e935e81e107cdc0c2361ec928baf3c4349e30cf @ 20200612, LLVM 10.0.0) + https://github.com/ispc/ispc/pull/1781/commits/da679386632e25bfb9d82de7f4cd9b9d88e8c937

**Comment by dbabokin** [comment]
The fail is reproducible with several other targets (but not all of them), including avx2-i32x8. For this target the test may be reduced further, by removing ``arr_29 = none(var_2);``.

Condition in ``if ((varying int8) var_14 < 0)`` may never be true, as var_14 is bool, i.e. it may have either value ``0`` or ``1``. So ideally the code should be completely removed. But this doesn't happen. Heres's the optput LLVM IR (which is correct one):
```llvm
; Function Attrs: nofree norecurse nounwind
define void @test() local_unnamed_addr #0 {
allocas:
  %var_14_load = load i8, i8* @var_14, align 1
  %0 = and i8 %var_14_load, 1
  %var_14_load_trunc_to_int8_broadcast_init = insertelement <8 x i8> undef, i8 %0, i32 0
  %1 = ashr <8 x i8> %var_14_load_trunc_to_int8_broadcast_init, <i8 7, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0, i8 0>
  %var_14_load_trunc_to_int8_broadcast.lobit = shufflevector <8 x i8> %1, <8 x i8> undef, <8 x i32> zeroinitializer
  %2 = icmp slt <8 x i8> %var_14_load_trunc_to_int8_broadcast.lobit, zeroinitializer
  %3 = bitcast <8 x i1> %2 to i8
  %cmp.i = icmp eq i8 %3, 0
  br i1 %cmp.i, label %safe_if_after_true, label %safe_if_run_true

safe_if_after_true:                               ; preds = %safe_if_run_true, %allocas
  ret void

safe_if_run_true:                                 ; preds = %allocas
  store i64 1, i64* @arr_68, align 8
  br label %safe_if_after_true
}
```

Note that doing ``and i8 %var_14_load, 1`` leads to at most only LBS bit set. So ``ashr 7`` leads to 0 in any case. So dead code may do a better job here.

Looks like code gen understands that this is a desad code and removes the code. But it does it incorrectly, leading to the following:
```asm
        vpmovmskb       eax, xmm0
        test    al, al
        je      .LBB2_2
.Ltmp4:
# %bb.1:                                # %safe_if_run_true
        .loc    2 9 9                   # example.ispc:9:9
        mov     qword ptr [rip + arr_68], 1
.Ltmp5:
.LBB2_2:                                # %safe_if_after_true
        #DEBUG_VALUE: __mask <- undef
        .loc    2 9 9 is_stmt 0         # example.ispc:9:9
        ret
```

This looks like LLVM code gen problem. It's reproducible with LLVM 10 and trunk.

**Comment by topperc** [comment]
Filed in llvm as https://bugs.llvm.org/show_bug.cgi?id=46585. Fixed in llvm commit 120c5f1057dc50229f73bc75bbabf4df6ee50fef

**Comment by dbabokin** [comment]
Fixed by #1812
